### DISCUSSION TITLE
Move all indexing/de-indexing work to queue

### DISCUSSION THREAD

**Discussion by janhenckens** [studioespresso/craft-scout#255] [Announcements]
This is the culmination of https://github.com/studioespresso/craft-scout/pull/248, https://github.com/studioespresso/craft-scout/pull/171 and https://github.com/studioespresso/craft-scout/discussions/247 by @aaronbushnell and @EpocSquadron.

Previously Scout would do the checking of wether or not something would need to be indexed or de-indexed inline and then do the indexing in a job if you had the `queue` config setting enabled. (de-indexing was always done inline)

The idea here is to move the checking to its own queue job, which has a significant impact on saving and editing.

You can try this by installing Scout like this:

    composer require studioespresso/craft-scout dev-feature/queue

This deprecates the `queue` setting and will thus no longer give you the option to run Scout without the queue.

Happy to hear your thoughts and feedback!

**Comment by aaronbushnell** [comment]
Hey @janhenckens — sorry for the delay! One thing I think needs to be added to the end of the `IndexElement.php` job is the following:

```php
// Only process the related elements if Scout is directed to
if (!Scout::$plugin->getSettings()->indexRelations) {
    return;
}

$element->getRelatedElements()->each(function (Element $relatedElement) {
    /* @var SearchableBehavior $relatedElement */
    $relatedElement->searchable(false);
});
```

Right now everything works as expected _except_ related element processing. These are currently not being updated during the Indexing job (when the option is enabled).

Here's why I _think_ it works on the current version of Scout (but not the beta):
- Previously, the `EVENT_BEFORE_DELETE_ELEMENT` and `EVENT_AFTER_DELETE_ELEMENT` were the exclusive owners of running this process
- Scout also wasn't checking if the element it was dealing with was searchable to begin with
- With that setup and without that check Scout could handle updates to related elements _within_ the `*_DELETE_ELEMENT` events

**Reply by aaronbushnell** [reply]
Sent a pull request to you for this, @janhenckens — https://github.com/studioespresso/craft-scout/pull/256

**Comment by aaronbushnell** [comment]
Just letting you know that we've been using this on a site since the 16th and so far so good, @janhenckens!

**Comment by aaronbushnell** [comment]
Hey @janhenckens — we're due to upgrade a site that's currently using 2.8.0-beta2 to Craft 4. Has that work been added to the Craft 4 version of this plugin? It _looks_ like it has but wanted to double check with you!

**Comment by janhenckens** [comment]
Hey @aaronbushnell, there's a Craft 4 version out but I'd have to double check to see if you the queue improvements you suggested/made in the beta for 3 are included in the current Craft 4 release. 

I'm at a conference for the rest of the week, I'll try to get back to you middle of next week!

**Comment by janhenckens** [comment]
@aaronbushnell The current release (`3.1.2`) has the changes from your PR so you should be good to upgrade to that.

**Reply by aaronbushnell** [reply]
Thanks so much, @janhenckens!
